### PR TITLE
Upgrade to Gradle 7.2 and enable Gradle modules on JDK EA build

### LINKED ISSUE #17081: Enable Gradle modules and tests on JDK 16+
Either by upgrading to Gradle 7 or by introducing toolchains and then reverting https://github.com/quarkusio/quarkus/pull/17053/commits/feef71aa76f01255edd7105ae84c215c989ab0c8 (⚠️ before removing that `jdk16-workarounds` profile from `build-parent/pom.xml`, please make sure there are no tests still being tagged with `failsOnJDK16`!).
Plus: revisit JDK 16 job `maven_args` in `.github/workflows/ci-actions-incremental.yml` (coming soon).

Relates to:
- https://github.com/gradle/gradle/issues/13481
- #16195

### LINKED ISSUE #19256: Upgrade Gradle wrapper to 7.0 or greater
### Description

Currently, Gradle 6.9 has some limitations as not being able to define [exclusiveContent](https://docs.gradle.org/current/userguide/declaring_repositories.html#declaring_content_exclusively_found_in_one_repository)

```
Gradle also lets you declare exclusive content that can only be found in one repository and it shouldn’t be searched for in any other.
```

This is especially useful to avoid "[A confusing dependency](https://autsoft.net/hu/a-confusing-dependency/)" than then later could finish in some security vulnerabilities. 

### Implementation ideas

Upgrade gradle distribution to something greater or equal to 7

https://github.com/quarkusio/quarkus/blob/main/devtools/gradle/gradle/wrapper/gradle-wrapper.properties#L3

### PR DESCRIPTION
**Pull request by glefloch**
This branch enables gradle modules build in `jdk-early-access-build` workflow.

close #17081 
close #19256

### CODE REVIEW

File: .github/workflows/jdk-early-access-build.yml
Comment (famod) [inline .github/workflows/jdk-early-access-build.yml:76]: Doesn't look like this issue was fixed so better leave those `action-surefire-report` steps disabled. Is this a leftover from a run in your fork?

**Review by famod — CHANGES_REQUESTED** [review]
So with this the job passed in your fork (with JDK `16` instead of `17-ea`)?

What about the disabled tests, see https://github.com/quarkusio/quarkus/pull/17053/commits/feef71aa76f01255edd7105ae84c215c989ab0c8? (please note that there was a wrong commit in the issue description, sorry)

**Comment by famod** [comment]
There is also something left here: https://github.com/quarkusio/quarkus/blob/main/.github/workflows/ci-actions-incremental.yml#L243 (remove `-pl '!devtools/gradle'`).

File: .github/workflows/jdk-early-access-build.yml
Comment (famod) [inline .github/workflows/jdk-early-access-build.yml:13]: Is PR still in WIP mode? I'm asking because the EA job moved on to 17-ea already and should not be rolled back like this.
What you can (and should) do though is trigger this job manually with `16` in your fork. No need to change the file.

**Review by famod — CHANGES_REQUESTED** [review]
It seems my previous comment still needs to be addressed as well: https://github.com/quarkusio/quarkus/pull/17102#issuecomment-836555114

Additionally, I think the other failing Gradle tests should be fixed first - if possible.

File: build-parent/pom.xml
Comment (famod) [inline build-parent/pom.xml:1063]: This slipped through in my recent review: Please don't remove this just yet, there are still other, non-Gradle tests that are skipped on Java 16 (until fixed).

File: .github/workflows/jdk-early-access-build.yml
Comment (glefloch) [reply .github/workflows/jdk-early-access-build.yml:13]: Oh no, I misunderstood your comment I will revert that

**Comment by famod** [comment]
Something must be wrong with my comment https://github.com/quarkusio/quarkus/pull/17102#issuecomment-836555114 because you keep missing it. 😉 

Btw, you sould rebase on `main` to receive test Gradle Kotlin test failures fix from your other PR.

File: .github/workflows/jdk-early-access-build.yml
Comment (famod) [reply .github/workflows/jdk-early-access-build.yml]: No worries, it's a monday!

File: integration-tests/devtools/src/test/java/io/quarkus/devtools/codestarts/quarkus/QuarkusCodestartBuildIT.java
Comment (famod) [inline integration-tests/devtools/src/test/java/io/quarkus/devtools/codestarts/quarkus/QuarkusCodestartBuildIT.java:60]: I don't think this will work just yet. We need Kotlin 1.5 to make this pass.

File: integration-tests/devtools/src/test/java/io/quarkus/devtools/codestarts/quarkus/QuarkusCodestartBuildIT.java
Comment (famod) [inline integration-tests/devtools/src/test/java/io/quarkus/devtools/codestarts/quarkus/QuarkusCodestartBuildIT.java:94]: Same as https://github.com/quarkusio/quarkus/pull/17102/files#r629535969, needs Kotlin 1.5.

**Comment by famod** [comment]
@glefloch I'm sorry for all those iterations, but it is all intermingled a bit, hence a bit complex.

**Comment by glefloch** [comment]
@famod no problem ;) I should be good now.

**Comment by famod** [comment]
I'm afraid this is not complete yet (have you really pushed your latest changes?):
- Still not covered: https://github.com/quarkusio/quarkus/pull/17102#issuecomment-836555114
  When I build `devtools/gradle` locally on Java 16 then `QuarkusPluginTest.shouldMakeQuarkusDevAndQuarkusBuildDependOnClassesTask()` fails with:
  ```
  org.gradle.api.GradleException: Could not inject synthetic classes.
  	at org.gradle.initialization.DefaultLegacyTypesSupport.injectEmptyInterfacesIntoClassLoader(DefaultLegacyTypesSupport.java:91)
  	at org.gradle.testfixtures.internal.ProjectBuilderImpl.getGlobalServices(ProjectBuilderImpl.java:183)
  	at org.gradle.testfixtures.internal.ProjectBuilderImpl.createProject(ProjectBuilderImpl.java:111)
  	at org.gradle.testfixtures.ProjectBuilder.build(ProjectBuilder.java:120)
  	at io.quarkus.gradle.QuarkusPluginTest.shouldMakeQuarkusDevAndQuarkusBuildDependOnClassesTask(QuarkusPluginTest.java:50)
  	[...]
  Caused by: java.lang.RuntimeException: java.lang.IllegalAccessException: module java.base does not open java.lang to unnamed module @2a79781f
  	at org.gradle.internal.classloader.ClassLoaderUtils$AbstractClassLoaderLookuper.invoke(ClassLoaderUtils.java:150)
  	at org.gradle.internal.classloader.ClassLoaderUtils$LookupClassDefiner.defineClass(ClassLoaderUtils.java:206)
  	at org.gradle.internal.classloader.ClassLoaderUtils.define(ClassLoaderUtils.java:77)
  	at org.gradle.initialization.DefaultLegacyTypesSupport.injectEmptyInterfacesIntoClassLoader(DefaultLegacyTypesSupport.java:88)
  	... 94 more
  Caused by: java.lang.IllegalAccessException: module java.base does not open java.lang to unnamed module @2a79781f
  	at java.base/java.lang.invoke.MethodHandles.privateLookupIn(MethodHandles.java:260)
  	at org.gradle.internal.classloader.ClassLoaderUtils$AbstractClassLoaderLookuper.getLookupForClassLoader(ClassLoaderUtils.java:159)
  	at org.gradle.internal.classloader.ClassLoaderUtils$AbstractClassLoaderLookuper.invoke(ClassLoaderUtils.java:146)
  	... 97 more
  ```
- All the tests I commented regarding " Kotlin 1.5" have to _stay tagged_ because we do not have Kotlin 1.5 yet! To make this clear: there are two problems in those tests: Gradle 6 _and_ Kotlin 1.4 not being fully Java 16 compatible.
- `CliTest.testGradleAddListRemove()` and `.testGradleBuild()` both fail in CI with compilation errors:
  ```
  Unsupported class file major version 60
  ```

TBH, it doesn't look like the Gradle 6.9 update changed that much in terms of Java 16 compatibility.
https://docs.gradle.org/6.9/release-notes.html
> Limited support for Java 16
> 
> This release does not support _running_ Gradle with JDK 16, but you can use Java toolchains to request Java 16 and compile your project.

**Comment by famod** [comment]
@glefloch any news on this one?

**Comment by glefloch** [comment]
I have been in a rush for last few weeks,  I should have time to look down to it this week.

**Comment by famod** [comment]
Cool! No stress, just wanted to prevent this from falling through the racks.

File: docs/src/main/asciidoc/gradle-tooling.adoc
Comment (glefloch) [inline docs/src/main/asciidoc/gradle-tooling.adoc:498]: @aloubyansky I added this piece of docs to explain why we have the `enforcedPlatform` and how to publish the application with it. WDYT ?

File: docs/src/main/asciidoc/gradle-tooling.adoc
Comment (gsmet) [inline docs/src/main/asciidoc/gradle-tooling.adoc:500]: Hmmm. Is `enforcedPlatform` still the way to go if it prevents people from publishing the app to Maven? Why did they do that?

Comment (glefloch) [reply docs/src/main/asciidoc/gradle-tooling.adoc:500]: well, in our case, `enforcedPlatform` answers our need as it overrides transitive dependency version. there is not a lot about why they did that, here is the issue that lead to that change: https://github.com/gradle/gradle/issues/15009
The main reason is 
````
enforced platforms shouldn't be used for published components because they behave like forced dependencies and leak to consumers. This can result in hard to diagnose dependency resolution errors. If you did this intentionally you can disable this check by adding 'enforced-platform' to the suppressed validations of the
````

Comment (gsmet) [reply docs/src/main/asciidoc/gradle-tooling.adoc:500]: OK, so we are not doing anything wrong I suppose. Thanks for the details!

**Review by gsmet — APPROVED** [review]
This looks good to me now but I will let @famod have the final say.

**Comment by famod** [comment]
I think `CliProjectGradleTest` can be re-enabled on Java 16. Do you want to do it @glefloch?

I'm planning on having a look at the remaining tests that are tagged with `failsOnJDK16` (mostly Kotlin), so I can also do it in that upcoming PR.

**Review by famod — CHANGES_REQUESTED** [review]
Great to see this coming to an end! 👍 

AFAICS, the only thing left (apart from that on disabled test I mentioned previously) is to remove `-pl '!devtools/gradle'` here: https://github.com/quarkusio/quarkus/blob/main/.github/workflows/ci-actions-incremental.yml#L253

**Comment by glefloch** [comment]
Right @famod, I enabled the cli test and the `ci-actions-incremental` job. Let's wait for CI :)

**Review by famod — APPROVED** [review]
Awesome!

File: docs/src/main/asciidoc/gradle-tooling.adoc
Comment (aloubyansky) [reply docs/src/main/asciidoc/gradle-tooling.adoc:498]: Thanks, looks good.

**Comment by famod** [comment]
@glefloch CI says:
```
Imports are not sorted in /home/runner/work/quarkus/quarkus/devtools/cli/src/test/java/io/quarkus/cli/CliProjectGradleTest.java
```

**Comment by famod** [comment]
@glefloch one test of `CliProjectGradleTest` failed on Windows, but on JDK 11, not 16!

Since this hasn't happend here before and since it seems to pass on `main`, I think it's a flaky test?

**Comment by famod** [comment]
- `MicroProfile TCKs Tests` failure is not related
- `Native Tests - Misc1` had a general problem with the Maven cache and I don't see a module of relevance in `"test-modules": "maven, jackson, jsonb, jsch, jgit, quartz, qute, consul-config, logging-min-level-unset, logging-min-level-set, simple with space"`

I have a CI run in progress in my fork that also incorporates this change and if that run does _not_ have the same `CliProjectGradleTest` failure then I'll just go ahead and merge this.

**Comment by famod** [comment]
Too bad, it failed in my fork as well. 😞

Looking at the trace, I'd bet it's a "some process is still locking files" windows problem:
```
org.opentest4j.AssertionFailedError: expected: <false> but was: <true>
	at org.junit.jupiter.api.AssertionUtils.fail(AssertionUtils.java:55)
	at org.junit.jupiter.api.AssertFalse.assertFalse(AssertFalse.java:40)
	at org.junit.jupiter.api.AssertFalse.assertFalse(AssertFalse.java:35)
	at org.junit.jupiter.api.Assertions.assertFalse(Assertions.java:210)
	at io.quarkus.cli.CliDriver.deleteDir(CliDriver.java:142)
	at io.quarkus.cli.CliProjectGradleTest.setupTestDirectories(CliProjectGradleTest.java:46)
```

**Comment by famod** [comment]
Hm, I cannot reproduce that test failure on my windows box. I'll push an update to Gradle 7.2 and see how it goes...

**Comment by famod** [comment]
`CliProjectGradleTest` failed again on Windows. 😒

**Comment by glefloch** [comment]
Yes, I think this is due to the daemon locking the directory. I will look if we can delete the projet after all tests.

**Comment by famod** [comment]
Yay, all green! 🎉

**Comment by gsmet** [comment]
This won't be backported for now and will be part of 2.3.

**Pull request merged**
